Support updates to payments

SteveIreland opened the pull request:
Okay, this is a rather trivial change that allows payments to be adjusted "post sale".

You simply go to Daily Sales to locate the invoice containing the payments to be adjusted and click on the "Update" icon.  You will be presented with the the Update form which now has a place to enter a payment adjustment.

Of course, there are a couple of problems I have with this entire process.  Payments are not transaction oriented by which I mean that if I add a payment transaction there isn't a record of the transaction written.  Instead the transaction is added to the collection of sales payments.  Since the key to the sales_payments table is the payment type there can only be one entry per payment type - so additional payments for a sale that all have the same payment type are accumulated into one record that reflects that payment type.

That's the negative side of this change.  The positive side is that you can now update the balance due for each sale if payments are received after the sale.

There are numerous improvements that can be made, but I wanted to offer this as both something that will allow some sites to temporarily "get by" and as a prototype for discussions about what future improvements might be necessary so that the job of payment tracking is done right.

SteveIreland commented:
The improvement I'd like to see made in the next release is to have the primary key to sales_payments be transaction_id (similar to the inventory table) and then have sale_id as a foreign key as well as additional columns for an "when entered" timestamp and a reference number for those who want to track check number, credit transaction number, etc.

WebShells commented:
@SteveIreland i submitted a pull request long time ago related to payments updates from the form, it was including an empty input box and update button once clicked it deduct the debt amount and add to cash amount, if you want to have a look... Definitely timestamp is needed for tracking, but i cant tell how to link it...

odiea commented:
Works good. The only suggestion I would have would be to show the current balance due on the form so the payment amount would be adjusted by that.

SteveIreland commented:
@odiea. Agreed. I started to add that but pulled back when I realized that the change would no longer be trivial. B cause we don’t track invoice totals I will need to basically run a complex SQL similar to what we do for the sales report to get the invoice total before I can start deducting payments to compute the balance due.

I believe if we move forward with enhancements to this effort I will add a ‘sales_stats’ to the system which will capture some of the summary data for a sale so that changes like this are much more efficient and easier

odiea commented:
@SteveIreland . I just added the following to the form.
```
<div class="form-group form-group-sm">
			<?php echo form_label($this->lang->line('sales_total'), 'sales_total', array('class'=>'control-label col-xs-3')); ?>
			<div class='col-xs-4'>
				<div class="input-group">
					<span class="input-group-addon input-sm"><b><?php echo $this->config->item('currency_symbol'); ?></span>
					<?php echo form_input(array(
							'name'=>'sales_total',
							'id'=>'sales_total',
							'class'=>'form-control input-sm', 'readonly'=>'true',
							'value'=>$sale_info['amount_due'])
							);?>
				</div>
			</div>
		</div>	
```

SteveIreland commented:
I've expanded this pull request. 

@odiea I went ahead and implemented the sales_stats and modified sales_payments to show where I'm headed with it.

I believe I might need to adapt one or more of the sales reports to the new structure of the sales_payment tables but so far everything seems to be functional.

jekkos commented on application/migrations/sqlscripts/3.3.0_paymenttracking.sql:
```diff
@@ -0,0 +1,38 @@
+  `payment_date` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  `reference_code` varchar(40) NOT NULL DEFAULT '',
+  PRIMARY KEY (`payment_id`),
+  KEY `paymnt_sale` (`sale_id`, `payment_type`)
```
minor typo here

SteveIreland replied:
done

jekkos commented on application/models/Sale.php:
```diff
@@ -112,6 +112,35 @@ public function get_info($sale_id)
+			//Get empty base object,
+			$stats_obj = new stdClass();
+
+			//Get all the fields from supplier table
```
comment is not completely correct, maybe just remove it

jekkos commented on application/migrations/sqlscripts/3.3.0_paymenttracking.sql:
```diff
@@ -0,0 +1,38 @@
+DROP TABLE IF EXISTS `ospos_sales_payments_backup`;
+
+CREATE TABLE `ospos_sales_stats` (
+  `sale_id` int(10) NOT NULL,
```
so this table will only be filled by new completed sales? Wouldn't there be a way to make it an sql view of the data instead? In that case it will always be uptodate and we need no code to manage it.

SteveIreland replied:
I thought I might get some push back on this table, however I'd like argue for the existence of this table although I'm certainly okay with dropping my use of the table for tracking payments.  In fact I'll go ahead and make that change.   

When I checked the function for `Sale->get_info()` the SQL overhead to retrieve information that will never change is enormous.  Once a sale is complete each of the amounts on the stats table won't change.  So calling get_info on sales (building two temporary tables and then joining multiple tables with group) seems like an unnecessary drain on resources when it can be replaced by a quick retrieve of two records (one from `sales` and one from `sales_stats`.

The sales_statistics table can also be used for user defined reporting that others, who are intimidated by the use of SQL to recalculate the sales totals,  might want to implement.

SteveIreland replied:
I just pushed my latest changes as described previously.  

I do believe the panel renders faster than before, but I'm biased so I even though it makes sense to me that the panel should render faster I'll let others be the judge.  It has to read the sales record, the customer record, the employee record and the set of payments before it can render so there were a couple of more reads than I first thought.   But still, it's no longer building temporary tables and reading the sales_items table, etc.

More testing is needed but I'm done for the weekend so I'll try to test some during the week.  I still have tested in a included tax scenario or with the destination based tax scenario.

jekkos replied:
Personally I'd like to stick to one plan of attack here. 

* Either we go for a stats table everywhere (and get rid of the temporary tables), and so we try to reuse the temporary tables for this case specifically. 

Reusing the temporary ones makes sense if we move it's creation to a piece of common code that can be called to find any aggregate value related to sales or payments.

Another argument for sticking with one approach is testability and correctness. Using one approach requires validation of only one resultset for further use, while mixing different approaches (temp and stats table) will lead to more complexity and bugs.

jekkos replied:
Also I think we can tackle performance issue with these temp tables by tweaking the time they can be kept in memory eg. (see some of the hints in the other perfomance PR here)

SteveIreland replied:
okay, I'll redo it.

SteveIreland replied:
done

SteveIreland replied:
Backed out the stats table.

jekkos commented on application/migrations/sqlscripts/3.3.0_paymenttracking.sql:
```diff
@@ -0,0 +1,38 @@
+-- This is to cleanup any orphaned tax migration tables
+
+RENAME TABLE `ospos_sales_payments` TO `ospos_sales_payments_backup`;
```
any specific reason for recreating the table here? If possible maybe just add the new columns on top of the old one?
Also I think you forgot to add the foreign key constraints so integrity checks are disabled here.

jekkos replied:
I guess it was done to fill the columns with default values. I'd just run an update query instead here. Also the default user id doesn't necessarily exist, so it could be a data integrity violation, best to make it nullable for the past.

SteveIreland replied:
Since I was changing the primary key my traditional approach is to populate the new version of the table using an insert so that the key can be auto assigned.

Good point on the user id, but I can actually populate it as part of the migration from the `sales` table.  That was something I hadn't yet thought about but will include that in the migration.  I feel it is important for audit purposes to be able to associate a user with each payment.

SteveIreland replied:
Done

jekkos commented on application/controllers/Sales.php:
```diff
@@ -1251,39 +1293,28 @@ public function save($sale_id = -1)
+		$payment_amount = $this->input->post('payment_amount_new');
+		$payment_type = $this->input->post('payment_type_new');
+
+		if($payment_type != '--' && $payment_amount <> 0)
```
don't know if we have a list of payment constant types but this '--' would then be the NONE constant.

SteveIreland replied:
done

jekkos commented on application/controllers/Sales.php:
```diff
@@ -1167,6 +1189,19 @@ public function edit($sale_id)
+
+		if($data['payment_amount_new'] != 0)
+		{
+			$data['balance_due'] = TRUE;
```
you could make this more concise (a oneliner), like `$data['balance_due'] = $data['payment_amount_new'] != 0`

SteveIreland replied:
done

SteveIreland commented:
I'm not sure if it's the result of changes with this pull request (I can't think of a reason why it would) but there seems to be an issue with the Payment Summary report and I will fix that tomorrow as part of this pull request regardless of what introduced the problem.

SteveIreland commented:
I found the issue with the payments report and fixed it as part of this pull request.  It wasn't caused by the changes in this pull request.  It turns out that the problem manifested itself in a "tax included" scenario.

Oh, and the issue was that the sales amount wasn't being included in the payments report. 

The pull request has been updated.  And if no one disapproves, since I've addressed all of the issues that @jekkos requested I will go ahead and merge it tomorrow night (around 24 hours from now) so that it has a wider audience for testing.

jekkos commented on application/migrations/sqlscripts/3.3.0_paymenttracking.sql:
```diff
@@ -0,0 +1,24 @@
+  `payment_date` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  `reference_code` varchar(40) NOT NULL DEFAULT '',
+  PRIMARY KEY (`payment_id`),
+  KEY `payment_sale` (`sale_id`, `payment_type`)
```
ok thanks for taking care of this, but we're still missing the foreign key constraints here

SteveIreland replied:
okay, I'll plug that in and test it before merging.

odiea commented:
@SteveIreland I don't know if anyone else gets this error. When I go to edit a new Invoice payment I receive the following error. It appears to be caused by an empty  $sale_info['amount_tendered']. It does not show if an additional payment has been made.
Severity: Warning --> A non-numeric value encountered C:\wamp64\www\my-master\application\controllers\Sales.php 1197 probably not the same line in Master.

SteveIreland commented:
Thanks. I will try to replicate the problem tonight and fix it

odiea commented:
Ok. Tested this am and no Issues. Thanks @SteveIreland

SteveIreland commented:
@odiea So the big question is did the changes fix the problems you found?

odiea commented:
So far. When I leave it in dev mode now I do not see any errors nor do I see any log errors..
